tools/convert_bigfile_gadget_format3.py

```diff
@@ -0,0 +1,271 @@
+        battr["UnitVelocity_in_cm_per_s"] = 100000.
+    #Some flags
+    battr["UsePeculiarVelocity"] = 0
+    #Missing a factor of 1/Hubble, but doesn't matter.
+    battr["RSDFactor"] = 1./hattr["Time"]**2
+    #Pass other keys through unchanged. We whitelist expected keys to avoid confusing Gadget.
+    hdfats = ["MassTable", "Time", "BoxSize", "Omega0", "OmegaLambda", "HubbleParam"]
+    for attr in hdfats:
+        battr[attr] = hattr[attr]
+
+def write_bf_segment(bf, hfile, startpart):
+    """Write the data arrays to an HDF5 file."""
+    #Open the file
+    with h5py.File(hfile,'r') as hdf5:
+        endpart = startpart + hdf5["Header"].attrs["NumPart_ThisFile"]
+        for ptype in range(6):
+            try:
+                htype = hdf5["PartType"+str(ptype)]
+            except KeyError:
+                continue
+            for hname in htype.keys():
+                block = names.get_bigfile_name(hname)
+                bname = "%d/%s" % (ptype, block)
+                harray = np.array(hdf5["PartType"+str(ptype)][hname])
+                #Beware this is not checked.
+                bf[bname].write(startpart[ptype], harray)
+        return endpart
+
+def create_big_file_arrays(bf, hfile):
+    """Pre-create all the big file arrays with the desired sizes."""
+    npart_tot = bf["Header"].attrs["TotNumPart"]
+    nfile = hfile["Header"].attrs["NumFilesPerSnapshot"]
+    for ptype in range(6):
+        try:
+            htype = hfile["PartType"+str(ptype)]
+        except KeyError:
+            continue
+        for hname in htype.keys():
+            hshape = np.shape(htype[hname])
+            dtype = htype[hname].dtype
+            block = names.get_bigfile_name(hname)
+            bname = "%d/%s" % (ptype, block)
+            try:
+                rows = hshape[1]
+            except IndexError:
+                rows = 1
+            bf.create(bname, dtype=(dtype, rows), size=npart_tot[ptype], Nfile=nfile)
+        #We need these (generally non-present) arrays but the simulation will start even if they are zero.
+        if ptype == 0 and "Generation" not in htype.keys():
+            genzero = np.zeros(npart_tot[ptype],dtype=np.uint8)
+            bf.create_from_array("0/Generation", genzero)
+        #MP-Gadget needs a mass array: Gadget-3 reads it from the header if all masses are the same.
+        if "Mass" not in htype.keys():
+            genzero = hfile["Header"].attrs["MassTable"][ptype] * np.ones(npart_tot[ptype],dtype=np.float32)
+            bf.create_from_array("%d/Mass" % ptype, genzero)
+
+
+def write_big_file(bfname, hdf5name):
+    """Find all the HDF5 files in the snapshot and merge them into a bigfile."""
+    #Find all the HDF5 snapshot set.
+    hdf5_files = glob.glob(hdf5name)
+    if len(hdf5_files) == 0:
+        hdf5_files = glob.glob(hdf5name+".*.hdf5")
+    elif os.path.isdir(hdf5_files[0]):
+        hdf5_files = glob.glob(os.path.join(hdf5name,"*_[0-9][0-9][0-9].*.hdf5"))
```

Comment:
Suggestion : sort file names to make sure the ordering is always consistent every time you run this.

Reply:
Yes! I always forget the output of glob isn't sorted. Well spotted!